### PR TITLE
feat(v2): composition syntax highlighting & live code editors

### LINKED ISSUE #1546: use composition style for prism-highlighting
If we do prism highlighting at render-time in React and not in the mdx loader, it would allow people to get access to code blocks as strings instead of html. This would make it easy for folks to add live code editors like [react-live](https://github.com/FormidableLabs/react-live).

We could keep the default prism setup too if we follow this guide in the MDX docs: https://mdxjs.com/guides/live-code#live-code. 

The benefits are pretty rad. I made a video of my setup for the formik v2 docs.

Kapture 2019-06-02 at 15 34 38

My swizzled `<DocsItem>` looks like this. As you can see, I am wrapping the DocContent with MDXProvider so I can override the code component with my playground.

```jsx
import React from 'react';
import Head from '@docusaurus/Head';
import DocPaginator from '@theme/DocPaginator';
import { MDXProvider } from '@mdx-js/react';
import styles from './styles.module.css';
import Playground from '../../components/Playground';


const Components = {};
Components.code = Playground;
Components.inlineCode = props => <code className="code" {...props} />;
Components.wrapper = props => <React.Fragment {...props} />;
Components.h1 = props => <h1 className="h1" {...props} />;
Components.h2 = props => <h2 className="h2" {...props} />;
Components.h3 = props => <h3 className="h3" {...props} />;
Components.h4 = props => <h4 className="h4" {...props} />;
Components.h5 = props => <h5 className="h5" {...props} />;
Components.h6 = props => <h6 className="h6" {...props} />;
Components.p = props => <p className="paragraph" {...props} />;
Components.ul = props => <ul className="ul" {...props} />;
Components.ol = props => <ol className="ol" {...props} />;
Components.li = props => <li className="li" {...props} />;
Components.a = props => <a className="link" {...props} />;
Components.blockquote = props => (
  <blockquote className="blockquote" {...props} />
);
Components.strong = props => <strong className="strong" {...props} />;
Components.pre = props => (
  <pre
    className="pre"
    style={{ backgroundColor: 'transparent', fontFamily: 'inherit' }}
    {...props}
  />
);

function Headings({ headings, isChild }) {
  if (!headings.length) return null;
  return (
    <ul className={isChild ? 'contents' : 'contents contents__left-border'}>
      {headings.map(heading => (
        <li key={heading.id}>
          <a href={`#${heading.id}`} className="contents__link">
            {heading.value}
          </a>
          <Headings isChild headings={heading.children} />
        </li>
      ))}
    </ul>
  );
}

function DocItem(props) {
  const { metadata, content: DocContent, docsMetadata } = props;

  return (
    <div className={styles.docBody}>
      <Head>
        {metadata && metadata.title && <title>{metadata.title}</title>}
      </Head>
      <div className="container margin-vert--lg">
        <div className="row">
          <div className="col col--8">
            <header>
              <h1 className="margin-bottom--lg">{metadata.title}</h1>
            </header>
            <article>
              <div className="markdown">
                <MDXProvider components={Components}>
                  <DocContent />
                </MDXProvider>
              </div>
            </article>
            <div className="margin-vert--lg" />
            <DocPaginator docsMetadata={docsMetadata} metadata={metadata} />
          </div>
          <div className="col col--3 col--offset-1">
            {DocContent.rightToc && <Headings headings={DocContent.rightToc} />}
          </div>
        </div>
      </div>
    </div>
  );
}

export default DocItem;
```

### PR DESCRIPTION
**Pull request by endiliey**
## Motivation

Close #1546 

Change how MDX component works, we can replace any component of mdx through mdx provider. 
I replaced `a` with `docusaurus/Link` to make internal links navigate without a page refresh and i replaced the syntax highlighting to be done via composition (which is recommended by https://mdxjs.com/guides/syntax-highlighting)

Add new live code editor, powered by react-live, enabled with this syntax. This make it still syntax highlightable 

```md
```js live
```

You can try it in netlify first for now, I added simple example at /docs/introduction (need to remove before merge)

Other notes
- 
- CodeBlock and playground is swizzle-able

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

use this https://gist.github.com/endiliey/b5f230c4cb6420641875612e25c3d4e1

live playground

ys

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/theme/components/Playground/index.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import {LiveProvider, LiveEditor, LiveError, LivePreview} from 'react-live';
+
+function Playground({children, theme, transformCode, ...props}) {
+  return (
+    <LiveProvider
+      code={children}
+      transformCode={transformCode || (code => `${code};`)}
+      theme={theme}
+      {...props}>
+      <div
```
Comment (yangshun) [inline packages/docusaurus-theme-classic/src/theme/components/Playground/index.js]: Must we use all these inline styles? Can we use CSS modules? Or do we want this part to be independent of webpack config? Inline styles are hard to override and I suspect users might want to do so.

File: packages/docusaurus-theme-classic/src/theme/Layout/index.js [hunk 2]
```diff
@@ -6,11 +6,13 @@
  */
 
 import React from 'react';
+import {MDXProvider} from '@mdx-js/react';
 import Head from '@docusaurus/Head';
 import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
 import withBaseUrl from '@docusaurus/withBaseUrl';
 import Navbar from '@theme/Navbar';
 import Footer from '@theme/Footer';
+import MDXComponents from '@theme/MDXComponents';
```
Comment (yangshun) [inline packages/docusaurus-theme-classic/src/theme/Layout/index.js]: This is a great idea 😄

File: packages/docusaurus-theme-classic/src/theme/components/Playground/index.js (hunk 1 above)
Comment (endiliey) [reply packages/docusaurus-theme-classic/src/theme/components/Playground/index.js]: yeah actually wanted to convert it to css modules. just havent done it yet

**Comment by endiliey** [comment]
Gonna merge this out first, yangshun will further customize it and tweak it with new PR

**Comment by jaredpalmer** [comment]
Not sure if this vibes with the swizzle philosophy, but it would be great to be able to specify component blocks from config.js with just `require.resolve(./myMDXComponents)`. The default export of that file then just gets past to MDXProvider. This is really all I wanted to customize

**Pull request merged**